src/nvim/tui/input.c

```diff
@@ -576,7 +609,34 @@ static void tinput_read_cb(Stream *stream, RBuffer *buf, size_t count_,
       }
     }
   } while (rbuffer_size(input->read_stream.buffer));
+}
+
+static void tinput_read_cb(Stream *stream, RBuffer *buf, size_t count_,
+                           void *data, bool eof)
+{
+  TermInput *input = data;
+
+  if (eof) {
+    loop_schedule_fast(&main_loop, event_create(tinput_done_event, 0));
+    return;
+  }
+
+  handle_raw_buffer(input, false);
   tinput_flush(input, true);
+
+  // An incomplete sequence was found. Leave it in the raw buffer and wait for
+  // the next input.
+  if (rbuffer_size(input->read_stream.buffer)) {
+    int ms = get_key_code_timeout();
+    // If ttimeout is not set, start the timer with a timeout of 0 to process
+    // the next input.
+    ms = ms > 0 ? ms : 0;
+    // Stop the current timer if already running
+    time_watcher_stop(&input->timer_handle);
+    time_watcher_start(&input->timer_handle, tinput_timer_cb, (uint32_t)ms, 0);
```

Comment:
so, scheduling the timer callback is the right thing for waiting the next input, is it right?

Reply:
I'm not sure if the answers below are appropriate for your question.

If there are more inputs or if the input has been split, the next call to `tinput_read_cb()` will process the remaining input in the buffer (`input->read_stream.buffer`). If there is no timer callback, if the input is a single `'\x1b'`, the remaining `'\x1b'` in the buffer will not be processed until the next input (As a result, it does not return to normal mode until the next input). To avoid that, we schedule a timer callback. That way, any remaining input will be processed when the timeout elapses.

I'm even sure that setting a timer callback with a timeout of 0 is correct. However, as far as I tried, it seems to work fine even if the input is split like `"\x1b [201"`, `"~"` (`"\x1b[201"` does not seem to be processed by the timer callback).